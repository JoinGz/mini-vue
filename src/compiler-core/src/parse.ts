import { NodeTypes } from './nodeTypes'

const openDelimiter = '{{'
const closeDelimiter = '}}'

const enum TagType {
  Start,
  End,
}

export function baseParse(str: string) {
  // 创建一个 当前文本的 上下文
  const context = createParserContext(str)
  // 解析children
  return createRoot(parseChildren(context, []))
}

function createParserContext(str: string) {
  return {
    sourse: str,
  }
}
function parseChildren(context: { sourse: string }, ancestors: string[]): any {
  const nodes = []

  while (!isEnd(context, ancestors)) {
    let node
    if (context.sourse.startsWith(openDelimiter)) {
      node = parseInterpolation(context)
    } else if (context.sourse[0] === '<') {
      if (/[A-Z]/i.test(context.sourse[1])) {
        node = parseElement(context, ancestors)
      }
    }

    if (!node) {
      node = parseText(context)
    }

    nodes.push(node)
  }

  return nodes
}

function createRoot(children: any) {
  return { children, type: NodeTypes.ROOT }
}

function parseInterpolation(context: { sourse: string }) {
  // msg = "{{message}}"

  const closeIndex = context.sourse.indexOf(
    closeDelimiter,
    openDelimiter.length
  )

  advanceBy(context, openDelimiter.length)

  const rowContentLength = closeIndex - openDelimiter.length

  const rowContent = parseTextData(context, rowContentLength)

  const content = rowContent.trim()

  advanceBy(context, closeDelimiter.length)

  // console.log(context.sourse)

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content: content,
    },
  }
}

function advanceBy(context: { sourse: string }, num: number) {
  return (context.sourse = context.sourse.slice(num))
}

function parseElement(context: { sourse: string }, ancestors: string[]): any {
  const element = parseTag(context, TagType.Start)

  ancestors.push(element.tag)
  element.children = parseChildren(context, ancestors)
  ancestors.pop()

  if (startsWithEndTagOpen(context, element.tag)) {
    parseTag(context, TagType.End)
  } else {
    throw new Error(`缺少结束标签:${element.tag}`)
  }

  console.log(context.sourse)

  return element
}

function parseTag(context: { sourse: string }, type: TagType): any {
  const tagMatch = context.sourse.match(/^<\/?([A-Z]*)/i)
  const tag = tagMatch![1]

  advanceBy(context, tagMatch![0].length) // 前进到<div>的前面的<div
  advanceBy(context, 1) // 前进到<div>的最后的>

  if (type === TagType.End) return

  return {
    type: NodeTypes.ELEMENT,
    tag,
  }
}

function parseText(context: { sourse: string }): any {
  let endIndex = context.sourse.length

  const endTokens = ['<', openDelimiter]

  for (let i = 0; i < endTokens.length; i++) {
    const tokens = endTokens[i]
    let index = context.sourse.indexOf(tokens)
    if (index !== -1 && endIndex > index) {
      endIndex = index
    }
  }

  const content = parseTextData(context, endIndex)

  return {
    type: NodeTypes.TEXT,
    content,
  }
}

function parseTextData(context: { sourse: string }, length: number) {
  const content = context.sourse.slice(0, length)

  advanceBy(context, length)

  // console.log(context.sourse);
  return content
}

function isEnd(context: { sourse: string }, ancestors: string[]) {
  if (context.sourse.startsWith('</')) {
    for (let i = ancestors.length - 1; i >= 0; i--) {
      const tag = ancestors[i]
      if (startsWithEndTagOpen(context, tag)) {
        return true
      }
    }
  }

  return !context.sourse
}

function startsWithEndTagOpen(context: { sourse: string }, tag: string) {
  return (
    context.sourse.startsWith('</') &&
    context.sourse.slice(2, 2 + tag.length).toLowerCase() === tag.toLowerCase()
  )
}
