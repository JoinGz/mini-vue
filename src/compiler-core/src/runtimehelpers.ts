import { Obj } from "../../../types/base";

export const TO_DISPLAY_STRING = Symbol("toDisplayString");

export const helperMapName: Obj = {
  [TO_DISPLAY_STRING]: "toDisplayString",
};
