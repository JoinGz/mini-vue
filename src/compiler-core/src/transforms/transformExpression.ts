import { NodeTypes } from "../nodeTypes";


export function transformExpression(ast: any) {
  if (ast.type === NodeTypes.INTERPOLATION) {
    processExpression(ast.content)
  }
}

function processExpression(ast: any): any {
  ast.content = `_ctx.${ast.content}`
}
