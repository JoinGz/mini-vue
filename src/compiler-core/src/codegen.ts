import { NodeTypes } from './nodeTypes'
import { helperMapName, TO_DISPLAY_STRING } from './runtimehelpers'


type codegenContext =  {
  code: string
  push(source: string): void
  helps: (str: any) => any
}

export function codegen(ast: any) {
  const functionName = `render`
  const codegenContext = createCodegenContext(ast)
  const { push } = codegenContext

  genFunctionPreamble(ast, codegenContext)

  push(`return `)
  const arg = ['_ctx']
  const argStr = arg.join(', ')

  push(`function`)
  push(` ${functionName} (${argStr}){`)

  genNode(ast.codegenNode, codegenContext)

  push('}')

  return {
    code: codegenContext.code,
  }
}

function createCodegenContext(ast: any) {
  const context = {
    code: '',
    push(source: string) {
      context.code += source
    },
    helps(key: string) {
      return `_${helperMapName[key]}`
    },
  }
  return context
}
function genNode(
  ast: any,
  codegenContext: codegenContext
) {
  switch (ast.type) {
    case NodeTypes.TEXT:
      genTextNode(ast, codegenContext)
      break
    case NodeTypes.INTERPOLATION:
      genInsertNode(ast, codegenContext)
      break
    case NodeTypes.SIMPLE_EXPRESSION:
      genExpression(ast, codegenContext)
      break

    default:
      break
  }
}

function genTextNode(
  ast: any,
  codegenContext: codegenContext
) {
  codegenContext.push(`return '${ast.content}'`)
}

function genFunctionPreamble(
  ast: any,
  codegenContext: codegenContext
) {
  // if (ast.children[0].type === NodeTypes.INTERPOLATION) {
    const vueBinging = `Vue`
    const { push } = codegenContext
    const { helps } = ast
    const astadd_ = (s: string) => `${helperMapName[s]}: _${helperMapName[s]}`

    if (helps.length) {
      push(`const { ${helps.map(astadd_).join(',')} } = ${vueBinging}`)
      push('\n')
    }
  // }
}

function genInsertNode(
  ast: any,
  codegenContext: {
    code: string
    push(source: string): void
    helps: (str: any) => any
  }
) {
  const { push, helps } = codegenContext
  push(`return ${helps(TO_DISPLAY_STRING)}(`)
  genNode(ast.content, codegenContext)
  push(`)`)
}

function genExpression(
  ast: any,
  codegenContext: codegenContext
) {
  const { push } = codegenContext
  push(`${ast.content}`)
}
