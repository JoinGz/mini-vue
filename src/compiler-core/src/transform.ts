import { Obj } from "../../../types/base";
import { NodeTypes } from "./nodeTypes";
import { TO_DISPLAY_STRING } from "./runtimehelpers";

export function transform(root: any, options: Obj = {}) {

  // 1. 深度遍历chileren
  // 2. 找到节点进行修改
  
  const context = createTransformContext(root, options)

  transformChildren(root, context)

  createRootCodegen(root)

  root.helps = [...context.helps.keys()]

  return root
}

function transformChildren(root: any, context: Obj) {
  
  const nodeTransforms = context.nodeTransforms || []
  
  for (let i = 0; i < nodeTransforms.length; i++) {
    const transform = nodeTransforms[i];
    transform(root)
  }
  
  traverseChildren(root, context);

}



function traverseChildren(root: any, context: any) {
  const children = root.children;


  switch (root.type) {
    case NodeTypes.INTERPOLATION:
        traverseExpression(root, context)
      break;
    case NodeTypes.ROOT:
    case NodeTypes.ELEMENT:
      traverseElement(children, context);
      break;
  
    default:
      break;
  }

}

function traverseElement(children: any, context: any) {
  for (let i = 0; i < children.length; i++) {
    const node = children[i];
    transformChildren(node, context);
  }
}

function createTransformContext(root: any, options: Obj) {
  const context =  {
    root: root,
    nodeTransforms: options.nodeTransforms || [],
    helps: new Map(),
    addHelp(key: string) {
      context.helps.set(key, 1)
    }
  }
  return context
}

function createRootCodegen(root: any) {
  root.codegenNode = root.children[0]
}

function traverseExpression(root: any, context: any) {
  context.addHelp(TO_DISPLAY_STRING)
}

