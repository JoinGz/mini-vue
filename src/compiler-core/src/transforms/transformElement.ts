import { NodeTypes } from "../nodeTypes";
import { CREATE_ELEMENT_VNODE } from "../runtimehelpers";


export function transformElement(ast: any, context: any) {


  return () => {
    if (ast.type === NodeTypes.ELEMENT) {
      context.addHelp(CREATE_ELEMENT_VNODE)
    }
    
  }


}


