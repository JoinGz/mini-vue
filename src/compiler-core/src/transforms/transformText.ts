import { NodeTypes } from '../nodeTypes'
import { CREATE_ELEMENT_VNODE } from '../runtimehelpers'

export function transformText(ast: any, context: any) {
  return () => {
    if (ast.type === NodeTypes.ELEMENT) {
      context.addHelp(CREATE_ELEMENT_VNODE)

      const { children } = ast

      let init

      for (let i = 0; i < children.length; i++) {
        const astNode = children[i]
        if (isText(astNode)) {
          for (let j = i + 1; j < children.length; j++) {
            const nextAstNode = children[j]
            if (isText(nextAstNode)) {
              if (!init) {
                init = children[i] = {
                  type: NodeTypes.COMPOUND_EXPRESSION,
                  children: [astNode],
                }
              }

              init.children.push(' + ')
              init.children.push(nextAstNode)
              children.splice(j, 1)
              j--
            } else {
              init = null
              break
            }
          }
        }
      }
    }
  }
}

function isText(ast: any) {
  return ast.type === NodeTypes.INTERPOLATION || ast.type === NodeTypes.TEXT
}
