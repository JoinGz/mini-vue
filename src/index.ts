export * from "./runtime-dom/index";

import * as runtimeDom from "./runtime-dom/index";


import { baseCompile } from './compiler-core/src/compiler'

import { createCompiler } from './runtime-core/index'


function createBaseCompile(template: string){
  const { code } = baseCompile(template)

  const render = new Function('Vue', code)(runtimeDom)

  return render
  
}

createCompiler(createBaseCompile)
