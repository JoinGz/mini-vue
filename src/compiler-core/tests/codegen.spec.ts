// code generation

import { codegen } from "../src/codegen"
import { baseParse } from "../src/parse"
import { transform } from "../src/transform"
import { transformExpression } from "../src/transforms/transformExpression"

describe('code-generation', () => {
  test('hi string', () => {
    const ast = baseParse('hi')

    transform(ast)

    const {code} = codegen(ast)

    

    expect(code).toMatchSnapshot()

  })
  
  test('expression', () => {
    const ast = baseParse('{{message}}')

    transform(ast,{nodeTransforms: [transformExpression]})

    const {code} = codegen(ast)

    

    expect(code).toMatchSnapshot()

  })
})