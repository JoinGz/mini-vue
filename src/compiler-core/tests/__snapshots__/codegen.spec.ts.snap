// Jest Snapshot v1, https://goo.gl/fbAQLP


exports[`code-generation hi string 1`] = `"return function render (_ctx){return 'hi'}"`;

exports[`code-generation expression 1`] = `
"const { toDisplayString: _toDisplayString } = Vue
return function render (_ctx){return _toDisplayString(_ctx.message)}"
`;
